src/Microsoft.TestPlatform.Client/TestPlatform.cs

```diff
@@ -303,21 +277,133 @@ private void AddExtensionAssembliesFromSource(IEnumerable<string> sources)
         {
-            var defaultExtensionPaths = fileHelper.EnumerateFiles(
-                extensionsFolder,
-                SearchOption.TopDirectoryOnly,
-                ".dll",
-                ".exe");
+            // Load default runtime providers
+            if ((strategy & TestAdapterLoadingStrategy.DefaultRuntimeProviders) == TestAdapterLoadingStrategy.DefaultRuntimeProviders)
+            {
+                defaultExtensionPaths = fileHelper
+                    .EnumerateFiles(extensionsFolder, SearchOption.TopDirectoryOnly, TestPlatformConstants.RunTimeEndsWithPattern)
+                    .Union(defaultExtensionPaths);
+            }
+
+            // Default extension loader
+            if (strategy == TestAdapterLoadingStrategy.Default
+                || (strategy & TestAdapterLoadingStrategy.ExtensionsDirectory) == TestAdapterLoadingStrategy.ExtensionsDirectory)
+            {
+                defaultExtensionPaths = fileHelper
+                    .EnumerateFiles(extensionsFolder, SearchOption.TopDirectoryOnly, ".dll", ".exe")
+                    .Union(defaultExtensionPaths);
+            }
+        }
+
+        TestPluginCache.Instance.DefaultExtensionPaths = defaultExtensionPaths.Distinct();
+    }
+
+    private static SearchOption GetSearchOption(TestAdapterLoadingStrategy strategy, SearchOption defaultStrategyOption) {
+        if (strategy == TestAdapterLoadingStrategy.Default) {
+            return defaultStrategyOption;
+        }
+
+        var searchOption = SearchOption.TopDirectoryOnly;
+        if ((strategy & TestAdapterLoadingStrategy.Recursive) == TestAdapterLoadingStrategy.Recursive)
+        {
+            searchOption = SearchOption.AllDirectories;
+        }
+
+        return searchOption;     
+    }
 
-            TestPluginCache.Instance.DefaultExtensionPaths = defaultExtensionPaths;
+    private static IEnumerable<string> ExpandTestAdapterPaths(string path, IFileHelper fileHelper, TestAdapterLoadingStrategy strategy)
+    {
+        var adapterPath = Path.GetFullPath(Environment.ExpandEnvironmentVariables(path));
+
+        // Default behavior is to only accept directories!
+        if (strategy == TestAdapterLoadingStrategy.Default)
+        {
+            return ExpandAdaptersWithDefaultStrategy(adapterPath, fileHelper);
+        }
+
+        var adapters = ExpandAdaptersWithExplicitStrategy(adapterPath, fileHelper, strategy);
+
+        return adapters.Distinct();
+    }
+
+    private static IEnumerable<string> ExpandAdaptersWithExplicitStrategy(string path, IFileHelper fileHelper, TestAdapterLoadingStrategy strategy)
+    {
+        if ((strategy & TestAdapterLoadingStrategy.Explicit) != TestAdapterLoadingStrategy.Explicit)
+        {
+            return Enumerable.Empty<string>();
         }
+
+        if (fileHelper.Exists(path))
+        {
+            return new[] { path };
+        }
+        else if (fileHelper.DirectoryExists(path))
```

Comment:
I would replace all the code from this point to the end of the method with a call to `TestAdapterLoadingStrategy.Explicit` which is doing exactly the same thing.

Reply:
My bad! Was referring to the following method (i.e. `ExpandAdaptersWithDefaultStrategy`).

Reply:
Ah, that's the legacy behavior. I kept it separate so we wont break it my mistake. I'll add a comment there.